### PR TITLE
Add support for `Option<*const T>`, `Option<*mut T>` and `NonNull<T>`

### PR DESCRIPTION
**Pull request by daxpedda**
I specifically didn't implement taking `NonNull<T>` as a parameter to avoid having to add a runtime check somewhere.

~~EDIT: Not sure whats going wrong here, but will leave it at draft until I figure it out.~~

### CODE REVIEW

File: src/convert/impls.rs [hunk 1]
```diff
@@ -223,6 +224,20 @@ impl<T> FromWasmAbi for *const T {
...
+        0
+    }
+}
+
+impl<T> OptionFromWasmAbi for *const T {
+    #[inline]
+    fn is_none(js: &u32) -> bool {
+        *js == 0
+    }
+}
```
Comment (Liamolucko) [inline src/convert/impls.rs:239]: This doesn't work: since `*const T` describes itself as a `u32`, as far as the CLI's concerned an `Option<*const T>` is the same as an `Option<u32>`, and `Option<u32>` doesn't use this representation. It directly implements `IntoWasmAbi`, without using `OptionIntoWasmAbi`, and works by adding an extra `u32` that represents 'is this some?'.

If you wanted to, you could add a new descriptor for pointers that does work this way, but I think doing it the same way as `Option<u32>` is better anyway since it's perfectly valid for a `*const T` to be null, and so that shouldn't be treated as some sentinel value.

On the other hand, null isn't a valid value for `NonNull<T>`, and so doing it this way for that would be fine (but it'd need a new descriptor).

File: examples/guide-supported-types-examples/src/non_null.rs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+pub unsafe fn take_pointer_by_value(x: Option<NonNull<u8>>) {
+    Box::from_raw(x.unwrap().as_ptr());
+}
+
+#[wasm_bindgen]
+pub fn return_pointer() -> Option<NonNull<u8>> {
+    NonNull::from(Box::new(42).leak())
```
Comment (Liamolucko) [inline examples/guide-supported-types-examples/src/non_null.rs:12]: I don't think this is valid: `Box::leak` can only be called as `Box::leak(x)`, not `x.leak()`, and this needs to either be wrapped in `Some` or not return an `Option`.

Also using `Box::leak` + `Box::from_raw` is a bit questionable, this should probably use `Box::into_raw` instead.

File: guide/src/reference/types/non-null.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# `*const T` and `*mut T`
```
Comment (Liamolucko) [inline guide/src/reference/types/non-null.md:1]:
```suggestion
# `NonNull<T>`
```

File: tests/wasm/simple.rs [hunk 4]
```diff
@@ -62,6 +64,20 @@ pub unsafe fn simple_raw_pointers_work(a: *mut u32, b: *const u8) -> *const u32
...
+pub unsafe fn simple_option_raw_pointers_work(
+    a: Option<*mut u32>,
+    b: Option<*const u8>,
+) -> Option<*const u32> {
+    *a.unwrap() = *b.unwrap() as u32;
+    a.map(|ptr| ptr as *const _)
+}
+
+#[wasm_bindgen]
+pub unsafe fn simple_option_nonnull_work(a: Option<NonNull<u32>>) -> Option<NonNull<u32>> {
```
Comment (Liamolucko) [inline tests/wasm/simple.rs:77]:
```suggestion
pub fn simple_option_nonnull_work(a: Option<NonNull<u32>>) -> Option<NonNull<u32>> {
```

Manipulating raw pointers isn't actually unsafe, just dereferencing them.

File: examples/guide-supported-types-examples/src/non_null.rs (hunk 2 above)
Comment (daxpedda) [reply examples/guide-supported-types-examples/src/non_null.rs:12]: Using [`Box::leak()` with `Box::from_raw()` is supported according to the documentation](https://doc.rust-lang.org/1.76.0/std/boxed/struct.Box.html#method.leak).

The reason why this is used is to get access to the infallible conversion to `NonNull`, compared to `NonNull::new()`. But in this case we return an `Option` anyway.

File: crates/cli-support/src/wit/incoming.rs [hunk 5]
```diff
@@ -155,6 +155,8 @@ impl InstructionBuilder<'_, '_> {
 
             // Largely synthetic and can't show up
             Descriptor::ClampedU8 => unreachable!(),
+
+            Descriptor::NonNull => self.number(AdapterType::NonNull, WasmVT::I32),
```
Comment (Liamolucko) [inline crates/cli-support/src/wit/incoming.rs:159]: It might be a good idea to add some kind of comment mentioning that this is unused right now due to only `Option<NonNull<T>>` implementing `FromWasmAbi`.

File: crates/cli-support/src/wit/standard.rs [hunk 6]
```diff
@@ -308,6 +309,8 @@ pub enum Instruction {
     OptionEnumFromI32 {
         hole: u32,
     },
+    InNonNullSentinel,
+    OutNonNullSentinel,
```
Comment (Liamolucko) [inline crates/cli-support/src/wit/standard.rs:313]: I'm not particularly a fan of these names; all the other `Instruction`s tend to use `From` and `To`, not `In` and `Out`, so they don't match very well. `OptionNonNullFromI32` and `I32FromOptionNonNull` would be more in line with everything else.

File: tests/wasm/simple.rs [hunk 7]
```diff
@@ -62,6 +66,35 @@ pub unsafe fn simple_raw_pointers_work(a: *mut u32, b: *const u8) -> *const u32
...
 }
 
+#[wasm_bindgen]
+pub unsafe fn simple_option_raw_pointers_work(
+    a: Option<*mut u32>,
+    b: Option<*const u8>,
+) -> Option<*const u32> {
+    *a.unwrap() = *b.unwrap() as u32;
+    a.map(|ptr| ptr as *const _)
+}
```
[the remark this answers is not in the document]
Comment (daxpedda) [reply tests/wasm/simple.rs]: Yeah, this was just checking if it compiles.

In any case, I hooked this and the old raw pointer tests up as well now.

File: crates/cli-support/src/js/binding.rs [hunk 8]
```diff
@@ -1217,6 +1217,18 @@ fn instruction(
...
+        Instruction::InNonNullSentinel => {
+            let val = js.pop();
+            js.cx.expose_is_like_none();
+            js.assert_optional_number(&val);
+            js.push(format!("isLikeNone({0}) ? 0 : {0}", val));
+        }
+
+        Instruction::OutNonNullSentinel => {
+            let val = js.pop();
+            js.push(format!("{0} === 0 ? undefined : {0}", val));
```
[the remark this answers is not in the document]
Comment (daxpedda) [reply crates/cli-support/src/js/binding.rs:1230]: I resolved it inside the `Instruction::WasmToInt` directly instead.

File: crates/cli-support/src/wit/incoming.rs (hunk 5 above)
Comment (daxpedda) [reply crates/cli-support/src/wit/incoming.rs]: I removed it and added an `unimplemented!()` instead.

File: crates/cli-support/src/js/binding.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Liamolucko) [reply crates/cli-support/src/js/binding.rs:1230]: `WasmToInt` isn't used in the optional case though, so this is still needed.

File: crates/cli-support/src/js/binding.rs [hunk 9]
```diff
@@ -1217,6 +1217,18 @@ fn instruction(
...
         }
+
+        Instruction::OptionNonNullFromI32 => {
+            let val = js.pop();
+            js.cx.expose_is_like_none();
+            js.assert_optional_number(&val);
+            js.push(format!("isLikeNone({0}) ? 0 : {0}", val));
+        }
+
+        Instruction::I32FromOptionNonNull => {
```
[the remark this answers is not in the document]
Comment (daxpedda) [reply crates/cli-support/src/js/binding.rs]: That's my bad!
I did double-check it but still mixed it up :sweat_smile: !

File: examples/guide-supported-types-examples/non_null.js
Comment (Liamolucko) [inline examples/guide-supported-types-examples/non_null.js:1]: This file's indentation is a bit messed up.

File: guide/src/SUMMARY.md [hunk 10]
```diff
@@ -53,7 +53,8 @@
     - [Exported Rust Types](./reference/types/exported-rust-types.md)
     - [`JsValue`](./reference/types/jsvalue.md)
     - [`Box<[T]>` and `Vec<T>`](./reference/types/boxed-slices.md)
-    - [`*const T` and `*mut T`](./reference/types/pointers.md)
+    - [`*const T`, `*mut T`](./reference/types/pointers.md)
```
Comment (Liamolucko) [inline guide/src/SUMMARY.md:56]: Why the change from 'and' to ','?

Comment (daxpedda) [reply guide/src/SUMMARY.md:56]: Thank you for the very careful review!

I previously had this and `NonNull` in the same file, so it was "`*const T`, `*mut T` and `NonNull`", I then split it into separate files and remove "and `NonNull`".

**Review by Liamolucko — APPROVED** [review]
LGTM!
